### PR TITLE
heterogeneous pool

### PR DESCRIPTION
**Pull request by wwanderley**
Pool proxy authentication like: https://github.com/oracle/node-oracledb/blob/master/doc/api.md#connpoolproxy

### CODE REVIEW

**Comment by tgulacsi** [comment]
This has nothing to do with proxy authentication.
That's already working with the "loginuser[proxyuser]" syntax.

I'd call this feature "heterogeneous pool" - please rename accordingly!

File: conn.go [hunk 1]
```diff
@@ -47,13 +47,19 @@ var _ = driver.ConnBeginTx((*conn)(nil))
 var _ = driver.ConnPrepareContext((*conn)(nil))
 var _ = driver.Pinger((*conn)(nil))
 
+type key int
+
+// ProxyUser specifies the proxy user in context
```
Comment (tgulacsi) [inline conn.go]: Not ProxyUser but CurrentUser.

File: conn.go [hunk 2]
```diff
@@ -47,13 +47,19 @@ var _ = driver.ConnBeginTx((*conn)(nil))
...
+const ProxyUser key = iota
+
 type conn struct {
 	sync.RWMutex
 	dpiConn        *C.dpiConn
 	connParams     ConnectionParams
 	inTransaction  bool
 	Client, Server VersionInfo
 	setTT          TraceTag
+	proxyUser      string
```
Comment (tgulacsi) [inline conn.go]: currentUser

File: conn.go [hunk 3]
```diff
@@ -488,3 +500,59 @@ type TraceTag struct {
 	// Action - specifies an action, such as an INSERT or UPDATE operation, in a module
 	Action string
 }
+
+func (c *conn) openProxyConnection(ctx context.Context) error {
```
Comment (tgulacsi) [inline conn.go]: ensureContextUser ?
The name should mirror the usage: this has to be called everywhere to change the connection according to the user requested in the context.

File: conn.go [hunk 4]
```diff
@@ -488,3 +500,59 @@ type TraceTag struct {
...
+	if c.dpiConn != nil {
+		if err := c.Close(); err != nil {
+			return driver.ErrBadConn
+		}
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	var connCreateParams C.dpiConnCreateParams
```
Comment (tgulacsi) [inline conn.go]: This should be refactored to share the common parts with openConn.

File: conn.go (hunk 2 above)
Comment (wwanderley) [reply conn.go]: renamed

File: conn.go (hunk 1 above)
Comment (wwanderley) [reply conn.go]: renamed

File: conn.go (hunk 3 above)
Comment (wwanderley) [reply conn.go]: renamed

File: conn.go (hunk 4 above)
Comment (wwanderley) [reply conn.go]: What's the best way to reuse openConn here? This function return a new conn struct.

Comment (wwanderley) [reply conn.go]: I created the acquireConn method to share commom parts of openConn with ensureContextUser.

**Comment by tgulacsi** [comment]
Thanks for the changes!

Can you add a test, too? 
Where is the password for the user?

**Comment by wwanderley** [comment]
> Thanks for the changes!
> 
> Can you add a test, too?
> Where is the password for the user?

Ok, I need to download a docker image to make some integration test.

**Comment by wwanderley** [comment]
@tgulacsi, I assume that the db user has privileges to create user. Is this ok?

**Pull request merged**